src/lib.rs

```diff
@@ -129,6 +189,38 @@ pub fn run(args: Args) {
                 .unwrap();
             buffer_writer.print(printer.get_mut()).unwrap();
         });
+
+    error_if_no_successful_query_parsing(&query_or_failure_by_language);
+}
+
+fn error_if_no_successful_query_parsing(
+    query_or_failure_by_language: &Mutex<HashMap<SupportedLanguageName, Option<Arc<Query>>>>,
+) {
+    let query_or_failure_by_language = query_or_failure_by_language.lock().unwrap();
+    if !query_or_failure_by_language
+        .values()
+        .any(|query| query.is_some())
+    {
+        fail("invalid query");
+    }
+}
+
+fn fail(message: &str) -> ! {
+    eprintln!("error: {message}");
+    process::exit(1);
+}
+
+fn get_and_cache_query_for_language(
+    query_source: &str,
+    query_or_failure_by_language: &Mutex<HashMap<SupportedLanguageName, Option<Arc<Query>>>>,
+    language: &dyn SupportedLanguage,
+) -> Option<Arc<Query>> {
+    query_or_failure_by_language
+        .lock()
+        .unwrap()
+        .entry(language.name())
+        .or_insert_with(|| maybe_get_query(query_source, language.language()).map(Arc::new))
```

Comment:
I thought a little bit about this in terms of performance - I don't know how slow query-parsing is (vs project file iteration) but there could be competition for this mutex especially "at the beginning" (if a bunch of project files get into `rayon`-world and get stuck waiting for the mutex to get access to the parsed query for their language)

But (in the "effectively single-language" scenario) they'd all have to in some sense wait for that query-parsing time anyway so doesn't seem terrible

If there are multiple supported languages being encountered then you could have parsing less-than-optimally parallelized I think, seems like then maybe having per-language mutexes would be the optimization there (rather than one mutex for n supported languages)?